### ISSUE TITLE
[FEAT EXTRACTION] Resampling issue

### ISSUE THREAD

**Issue by Jonas1312** [AIM-Harvard/pyradiomics#686]
Hi,

I would like to extract some radiomic features from images and masks with different shapes and spacings...

Initially, masks and images have the same corresponding shapes and spacing. But before sending an image to Pyradiomics, it goes through a pipeline that applies several transformations including a bilinear resampling on the x and y dimensions. The tumour mask/segmentation is not processed through that pipeline (thus, no resampling is applied).

My idea was to let Pyradiomics resample the mask on (x, y, z) axes, and resample the input image only on the axial plane (z-axis):
- input image with shape (311, 311, 300) and spacing (1, 1, 1)
- mask with shape (512, 512, 300) and spacing (0.607421875, 0.607421875, 1.0).

It works on most of my images except on one where Pyradiomics raises an error:
```text
ValueError: mask has too few dimensions (number of dimensions 2, minimum required 3)
```

I wrote a simple script below which reproduces the issue. The script shows that when i let Pyradiomics resample both the image and the mask, it works. But when i resample the image myself (x and y axis), while using the same mask, Pyradiomics fails.

Thanks!

**PyRadiomics configuration**
```python
RadiomicsFeatureExtractor(
    minimumROIDimensions=3, resampledPixelSpacing=[1.0, 1.0, 1.0], interpolator=2
)
```

**PyRadiomics log file**
```text
INFO:radiomics.featureextractor: No valid config parameter, using defaults: {'minimumROIDimensions': 2, 'minimumROISize': None, 'normalize': False, 'normalizeScale': 1, 'removeOutliers': None, 'resampledPixelSpacing': None, 'interpolator': 'sitkBSpline', 'preCrop': False, 'padDistance': 5, 'distances': [1], 'force2D': False, 'force2Ddimension': 0, 'resegmentRange': None, 'label': 1, 'additionalInfo': True}
INFO:radiomics.featureextractor: Enabled image types: {'Original': {}}
INFO:radiomics.featureextractor: Enabled features: {'firstorder': [], 'glcm': [], 'gldm': [], 'glrlm': [], 'glszm': [], 'ngtdm': [], 'shape': []}
INFO:radiomics.featureextractor: Applying custom setting overrides: {'minimumROIDimensions': 3, 'resampledPixelSpacing': [1.0, 1.0, 1.0], 'interpolator': 2}
DEBUG:radiomics.featureextractor: Settings: {'minimumROIDimensions': 3, 'minimumROISize': None, 'normalize': False, 'normalizeScale': 1, 'removeOutliers': None, 'resampledPixelSpacing': [1.0, 1.0, 1.0], 'interpolator': 2, 'preCrop': False, 'padDistance': 5, 'distances': [1], 'force2D': False, 'force2Ddimension': 0, 'resegmentRange': None, 'label': 1, 'additionalInfo': True}
INFO:radiomics.featureextractor: Calculating features with label: 1
DEBUG:radiomics.featureextractor: Enabled images types: {'Original': {}}
DEBUG:radiomics.featureextractor: Enabled features: {'firstorder': [], 'glcm': [], 'gldm': [], 'glrlm': [], 'glszm': [], 'ngtdm': [], 'shape': []}
DEBUG:radiomics.featureextractor: Current settings: {'minimumROIDimensions': 3, 'minimumROISize': None, 'normalize': False, 'normalizeScale': 1, 'removeOutliers': None, 'resampledPixelSpacing': [1.0, 1.0, 1.0], 'interpolator': 2, 'preCrop': False, 'padDistance': 5, 'distances': [1], 'force2D': False, 'force2Ddimension': 0, 'resegmentRange': None, 'label': 1, 'additionalInfo': True}
INFO:radiomics.featureextractor: Loading image and mask
DEBUG:radiomics.imageoperations: Force casting mask to UInt32 to ensure correct datatype.
DEBUG:radiomics.imageoperations: Resampling image and mask
DEBUG:radiomics.imageoperations: Where resampled spacing is set to 0, set it to the original spacing (mask)
DEBUG:radiomics.imageoperations: Checking ROI validity
DEBUG:radiomics.imageoperations: Checking if label 1 is persent in the mask
DEBUG:radiomics.imageoperations: Comparing physical space of bounding box to physical space of image
DEBUG:radiomics.imageoperations: ROI bounds (image coordinate space): ((62.5, 212.5, 130.5), (65.5, 215.5, 133.5))
DEBUG:radiomics.imageoperations: ROI valid, calculating resampling grid
DEBUG:radiomics.imageoperations: Comparing resampled spacing to original spacing (image
INFO:radiomics.imageoperations: Applying resampling from spacing [0.60742188 0.60742188 1.        ] and size [512 512 300] to spacing [1. 1. 1.] and size [14, 13, 15]
DEBUG:radiomics.imageoperations: Resampling image
DEBUG:radiomics.imageoperations: Resampling mask
DEBUG:radiomics.imageoperations: Checking mask with label 1
DEBUG:radiomics.imageoperations: Calculating bounding box
DEBUG:radiomics.imageoperations: Checking minimum number of dimensions requirements (3)
DEBUG:radiomics.featureextractor: Image and Mask loaded and valid, starting extraction
DEBUG:radiomics.imageoperations: Cropping to size [2 2 3]
INFO:radiomics.featureextractor: Computing shape
DEBUG:radiomics.shape: Initializing feature class
DEBUG:radiomics.shape: Padding the mask with 0s
DEBUG:radiomics.shape: Pre-calculate Volume, Surface Area and Eigenvalues
DEBUG:radiomics.shape: Shape feature class initialized
DEBUG:radiomics.shape: Calculating features
DEBUG:radiomics.featureextractor: Creating image type iterator
INFO:radiomics.featureextractor: Adding image type "Original" with custom settings: {}
DEBUG:radiomics.featureextractor: Extracting features
DEBUG:radiomics.imageoperations: Yielding original image
INFO:radiomics.featureextractor: Calculating features for original image
DEBUG:radiomics.imageoperations: Cropping to size [2 2 3]
INFO:radiomics.featureextractor: Computing firstorder
DEBUG:radiomics.firstorder: Initializing feature class
DEBUG:radiomics.imageoperations: Discretizing gray levels inside ROI
DEBUG:radiomics.imageoperations: Calculated 29 bins for bin width 25 with edges: [ 550  575  600  625  650  675  700  725  750  775  800  825  850  875
  900  925  950  975 1000 1025 1050 1075 1100 1125 1150 1175 1200 1225
 1250 1275])
DEBUG:radiomics.firstorder: First order feature class initialized
DEBUG:radiomics.firstorder: Calculating features
INFO:radiomics.featureextractor: Computing glcm
DEBUG:radiomics.glcm: Initializing feature class
DEBUG:radiomics.imageoperations: Discretizing gray levels inside ROI
DEBUG:radiomics.imageoperations: Calculated 29 bins for bin width 25 with edges: [ 550  575  600  625  650  675  700  725  750  775  800  825  850  875
  900  925  950  975 1000 1025 1050 1075 1100 1125 1150 1175 1200 1225
 1250 1275])
DEBUG:radiomics.glcm: Calculating GLCM matrix in C
DEBUG:radiomics.glcm: Process calculated matrix
DEBUG:radiomics.glcm: Create symmetrical matrix
DEBUG:radiomics.glcm: Deleting 4 empty angles:
[[ 1  1  1]
 [ 1  1 -1]
 [ 1 -1 -1]
 [ 0  1  1]]
DEBUG:radiomics.glcm: Calculating GLCM coefficients
DEBUG:radiomics.glcm: GLCM feature class initialized, calculated GLCM with shape (1, 5, 5, 9)
DEBUG:radiomics.glcm: Calculating features
WARNING:radiomics.glcm: GLCM is symmetrical, therefore Sum Average = 2 * Joint Average, only 1 needs to be calculated
INFO:radiomics.featureextractor: Computing gldm
DEBUG:radiomics.gldm: Initializing feature class
DEBUG:radiomics.imageoperations: Discretizing gray levels inside ROI
DEBUG:radiomics.imageoperations: Calculated 29 bins for bin width 25 with edges: [ 550  575  600  625  650  675  700  725  750  775  800  825  850  875
  900  925  950  975 1000 1025 1050 1075 1100 1125 1150 1175 1200 1225
 1250 1275])
DEBUG:radiomics.gldm: Calculating GLDM matrix in C
DEBUG:radiomics.gldm: Feature class initialized, calculated GLDM with shape (1, 5, 1)
DEBUG:radiomics.gldm: Calculating features
INFO:radiomics.featureextractor: Computing glrlm
DEBUG:radiomics.glrlm: Initializing feature class
DEBUG:radiomics.imageoperations: Discretizing gray levels inside ROI
DEBUG:radiomics.imageoperations: Calculated 29 bins for bin width 25 with edges: [ 550  575  600  625  650  675  700  725  750  775  800  825  850  875
  900  925  950  975 1000 1025 1050 1075 1100 1125 1150 1175 1200 1225
 1250 1275])
DEBUG:radiomics.glrlm: Calculating GLRLM matrix in C
DEBUG:radiomics.glrlm: Process calculated matrix
DEBUG:radiomics.glrlm: Deleting 4 empty angles:
[[ 1  1  1]
 [ 1  1 -1]
 [ 1 -1 -1]
 [ 0  1  1]]
DEBUG:radiomics.glrlm: Calculating GLRLM coefficients
DEBUG:radiomics.glrlm: GLRLM feature class initialized, calculated GLRLM with shape (1, 5, 1, 9)
DEBUG:radiomics.glrlm: Calculating features
INFO:radiomics.featureextractor: Computing glszm
DEBUG:radiomics.glszm: Initializing feature class
DEBUG:radiomics.imageoperations: Discretizing gray levels inside ROI
DEBUG:radiomics.imageoperations: Calculated 29 bins for bin width 25 with edges: [ 550  575  600  625  650  675  700  725  750  775  800  825  850  875
  900  925  950  975 1000 1025 1050 1075 1100 1125 1150 1175 1200 1225
 1250 1275])
DEBUG:radiomics.glszm: Calculating GLSZM matrix in C
DEBUG:radiomics.glszm: Calculating GLSZM coefficients
DEBUG:radiomics.glszm: GLSZM feature class initialized, calculated GLSZM with shape (1, 5, 1)
DEBUG:radiomics.glszm: Calculating features
INFO:radiomics.featureextractor: Computing ngtdm
DEBUG:radiomics.ngtdm: Initializing feature class
DEBUG:radiomics.imageoperations: Discretizing gray levels inside ROI
DEBUG:radiomics.imageoperations: Calculated 29 bins for bin width 25 with edges: [ 550  575  600  625  650  675  700  725  750  775  800  825  850  875
  900  925  950  975 1000 1025 1050 1075 1100 1125 1150 1175 1200 1225
 1250 1275])
DEBUG:radiomics.ngtdm: Calculating features
DEBUG:radiomics.featureextractor: Features extracted
INFO:radiomics.featureextractor: Calculating features with label: 1
DEBUG:radiomics.featureextractor: Enabled images types: {'Original': {}}
DEBUG:radiomics.featureextractor: Enabled features: {'firstorder': [], 'glcm': [], 'gldm': [], 'glrlm': [], 'glszm': [], 'ngtdm': [], 'shape': []}
DEBUG:radiomics.featureextractor: Current settings: {'minimumROIDimensions': 3, 'minimumROISize': None, 'normalize': False, 'normalizeScale': 1, 'removeOutliers': None, 'resampledPixelSpacing': [1.0, 1.0, 1.0], 'interpolator': 2, 'preCrop': False, 'padDistance': 5, 'distances': [1], 'force2D': False, 'force2Ddimension': 0, 'resegmentRange': None, 'label': 1, 'additionalInfo': True}
INFO:radiomics.featureextractor: Loading image and mask
DEBUG:radiomics.imageoperations: Force casting mask to UInt32 to ensure correct datatype.
DEBUG:radiomics.imageoperations: Resampling image and mask
DEBUG:radiomics.imageoperations: Where resampled spacing is set to 0, set it to the original spacing (mask)
DEBUG:radiomics.imageoperations: Checking ROI validity
DEBUG:radiomics.imageoperations: Checking if label 1 is persent in the mask
DEBUG:radiomics.imageoperations: Comparing physical space of bounding box to physical space of image
DEBUG:radiomics.imageoperations: ROI bounds (image coordinate space): ((37.9638671875, 129.0771484375, 130.5), (39.7861328125, 130.8994140625, 133.5))
DEBUG:radiomics.imageoperations: ROI valid, calculating resampling grid
DEBUG:radiomics.imageoperations: Comparing resampled spacing to original spacing (image
INFO:radiomics.imageoperations: New spacing equal to original image spacing, just resampling the mask
DEBUG:radiomics.imageoperations: Cropping to size [2 1 3]
DEBUG:radiomics.imageoperations: Checking mask with label 1
DEBUG:radiomics.imageoperations: Calculating bounding box
DEBUG:radiomics.imageoperations: Checking minimum number of dimensions requirements (3)
```

**To Reproduce**
```python
import numpy as np
import SimpleITK as sitk
from radiomics import featureextractor


def print_sitk_info(sitk_image, name):
    print(f"--------------- {name}:")
    print("Shape: ", sitk_image.GetSize())
    print("Spacing: ", sitk_image.GetSpacing())
    print("Direction: ", sitk_image.GetDirection())
    print("Origin: ", sitk_image.GetOrigin())


def main():
    # create first (fake) ct scan
    # x and y not isotropic
    ct_scan_numpy = np.random.randint(1337, size=(512, 512, 300), dtype=np.int32)
    ct_scan_numpy = np.transpose(ct_scan_numpy, axes=(2, 0, 1))  # channel first
    ct_scan_sitk = sitk.GetImageFromArray(ct_scan_numpy)  # numpy to sitk
    ct_scan_sitk.SetSpacing((0.607421875, 0.607421875, 1.0))
    ct_scan_sitk_1 = ct_scan_sitk

    # create 2nd ct scan
    # x and y axes already resampled
    ct_scan_numpy = np.random.randint(1337, size=(311, 311, 300), dtype=np.int32)
    ct_scan_numpy = np.transpose(ct_scan_numpy, axes=(2, 0, 1))  # channel first
    ct_scan_sitk = sitk.GetImageFromArray(ct_scan_numpy)  # numpy to sitk
    ct_scan_sitk.SetSpacing((1.0, 1.0, 1.0))
    ct_scan_sitk_2 = ct_scan_sitk

    # load mask (not resampled)
    nod_mask_numpy = np.zeros((512, 512, 300), dtype=np.int32)
    # fmt: off
    nodule_coords = (
        np.array([213, 213, 213, 214, 214, 214, 214, 214, 214, 214, 214, 215, 215,
        215, 215, 215, 215], dtype=np.int64),
        np.array([64, 64, 65, 63, 63, 64, 64, 64, 65, 65, 65, 63, 64, 64, 64, 65, 65],
        dtype=np.int64),
        np.array([131, 132, 132, 131, 132, 131, 132, 133, 131, 132, 133, 132, 131,
        132, 133, 132, 133], dtype=np.int64)
    )
    # fmt: on
    nod_mask_numpy[nodule_coords] = 1
    nod_mask_numpy = np.transpose(nod_mask_numpy, axes=(2, 0, 1))  # channel first
    mask_sitk = sitk.GetImageFromArray(nod_mask_numpy)  # numpy to sitk
    mask_sitk.SetSpacing((0.607421875, 0.607421875, 1.0))
    mask_sitk = sitk.Cast(mask_sitk, sitk.sitkInt32)  # bool to int

    # print infos
    print_sitk_info(ct_scan_sitk_1, name="ct_scan_sitk_1")
    print_sitk_info(ct_scan_sitk_2, name="ct_scan_sitk_2")
    print_sitk_info(mask_sitk, name="mask")

    # init RadiomicsFeatureExtractor
    extractor = featureextractor.RadiomicsFeatureExtractor(
        minimumROIDimensions=3, resampledPixelSpacing=[1.0, 1.0, 1.0], interpolator=2
    )

    rad_features = extractor.execute(ct_scan_sitk_1, mask_sitk)  # success
    rad_features = extractor.execute(ct_scan_sitk_2, mask_sitk)  # fail


if __name__ == "__main__":
    main()
```

**Version (please complete the following information):**
 - OS: Windows 20H2
 - Python version: 3.8.8
 - PyRadiomics version: 3.0.1

**Comment by Jonas1312** [comment]
In the first case (image and mask not resampled), Pyradiomics crops to size [2 2 3]. For the 2nd case (image resampled, not mask), Pyradiomics crops to size [2 1 3], which raises the "mask has too few dimensions" error.

That's weird because it's the same mask in both cases...

**Comment by JoostJM** [comment]
**tldr**: It's not an error, but PyRadiomics doing what you asked to do: skip masks that are too small (in this case, < 3 dimensions). The difference between cases is most likely due to alignment differences in resampling grids between your pipeline and PyRadiomics. Coupled with a very small mask, this results in a resampled mask with size = 1 in Y direction (= 2 dimensions or single slice mask).

**Full answer**:

The 'error' is due to the fact that the mask is upsampled to spacing [1, 1, 1]. In your specific case the mask is very small (only 12 voxels of bounding box in the first case, only 6 in the second). The difference in size after resampling is probably due to differences in alignment of the resampling grid in your pipeline vs the implementation in PyRadiomics (in the latter, the corner of the first voxel is aligned).

These small differences in alignment can cause the resampling of the mask to result in a somewhat smaller or larger mask, in this case it's a bit smaller in the PyRadiomics resampling after application of your pipeline. Given the fact that the mask was very small to begin with, the result was a size of 1 in the Y dimension. As you have set "minimum ROI dimensions" to 3, PyRadiomics the fails as the mask is considered "invalid".

By default, PyRadiomics allows for 2D masks. Is there a specific reason you force your masks to be at least 3D-masks? Not setting the minimumROI dimensions does not change the extraction, it just allows masks which represent a single slice segmentation to be used for extraction.

That said, I'd still want to point out that in both cases, the validity of the extracted features is questionable, as with such a small mask, the features are based on very little information and one can wonder to what extent the contain valid information regarding the true underlying texture.
